### PR TITLE
[HLSL][clang] Move hlsl_wave_get_lane_index to EmitHLSLBuiltinExpr

### LINKED ISSUE #87109: Move __builtin_hlsl_wave_get_lane_index to EmitHLSLBuiltinExpr
`Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID, const CallExpr *E)`

is where all the other hlsl builtins live.

_Originally posted by @farzonl in https://github.com/llvm/llvm-project/pull/80680#discussion_r1544449419_

### PR DESCRIPTION
**Pull request by marcauberer**
Resolves #87109

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -18317,6 +18307,14 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
         /*ReturnType=*/Op0->getType(), Intrinsic::dx_rsqrt,
         ArrayRef<Value *>{Op0}, nullptr, "dx.rsqrt");
   }
+  case Builtin::BI__builtin_hlsl_wave_get_lane_index: {
+    auto *CI = EmitRuntimeCall(CGM.CreateRuntimeFunction(
+        llvm::FunctionType::get(IntTy, {}, false), "__hlsl_wave_get_lane_index",
+        {}, false, true));
+    if (getTarget().getTriple().isSPIRVLogical())
+      CI = dyn_cast<CallInst>(addControlledConvergenceToken(CI));
+    return RValue::get(CI);
```
Comment (farzonl) [inline clang/lib/CodeGen/CGBuiltin.cpp:18316]: this isn't going to work  we don't return RValues from this function that happens later see: https://github.com/llvm/llvm-project/blob/a8b0ecd2605ff23f495a8af64e06c35f86834e54/clang/lib/CodeGen/CGBuiltin.cpp#L6192C3-L6193C27

Comment (farzonl) [reply clang/lib/CodeGen/CGBuiltin.cpp:18316]: Please build the project and run the hlsl lit tests

**Comment by farzonl** [comment]
Hi @marcauberer  The change looks good. I'm 99% sure this won't have an effect on the SPIRV backend, however there is still that 1% chance it does.

The tests I want to check landed in this PR
https://github.com/llvm/llvm-project/pull/85979/files 

For convenience this is what you can pass to `llvm-lit`:
- llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveGetLaneIndex.ll  (really this one is the one im concerned about)
- llvm/test/CodeGen/SPIRV/scfg-add-pre-headers.ll 
- llvm/test/CodeGen/SPIRV/transcoding/spirv-private-array-initialization.ll


Since you aren't changing anything in the SPIRV backend the github action didn't trigger. So if you could do one of two things   after I approve, I would appreciate it,

Option 1: build spirv backend and test the above test cases
Should be a straight forward build
```
cmake -B <spirv-debug-build-dir> -G Ninja -S llvm -DLLVM_ENABLE_PROJECTS="llvm;clang" -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_BUILD_TYPE=Debug -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=SPIRV -DLLVM_INCLUDE_SPIRV_TOOLS_TESTS=ON
```
Option 2: watch or trigger the SPIR_V Test github actions
https://github.com/llvm/llvm-project/actions/workflows/spirv-tests.yml
make sure the tests pass if they don't and it is one of the tests listed above you might have to revert your change.

**Review by farzonl — APPROVED** [review]

**Comment by farzonl** [comment]
minor nit pick, you don't have to do this one: could you make the HLSL prefix first. It helps with some bookeeping we are doing.

**Comment by marcauberer** [comment]
@farzonl All three tests pass 👍🏽 

```sh
$ ./bin/llvm-lit ../llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveGetLaneIndex.ll ../llvm/test/CodeGen/SPIRV/scfg-add-pre-headers.ll ../llvm/test/CodeGen/SPIRV/transcoding/spirv-private-array-initialization.ll
-- Testing: 3 tests, 3 workers --
PASS: LLVM :: CodeGen/SPIRV/scfg-add-pre-headers.ll (1 of 3)
PASS: LLVM :: CodeGen/SPIRV/transcoding/spirv-private-array-initialization.ll (2 of 3)
PASS: LLVM :: CodeGen/SPIRV/hlsl-intrinsics/WaveGetLaneIndex.ll (3 of 3)

Testing Time: 0.04s

Total Discovered Tests: 3
  Passed: 3 (100.00%)
```

**Comment by marcauberer** [comment]
Workflow run also succeeded: https://github.com/llvm/llvm-project/actions/runs/8493247308

**Pull request merged**